ipa-core/src/protocol/dp/distributions.rs

```diff
@@ -63,44 +61,139 @@ impl Distribution<f64> for RoundedBoxMuller {
         self.inner.sample(rng).round()
     }
 }
-
 impl From<BoxMuller> for RoundedBoxMuller {
     fn from(value: BoxMuller) -> Self {
         Self { inner: value }
     }
 }
 
+///  What follows is the implementation of sampling from the Truncated Double Geometric distribution which in the
+///  process samples from Double Geometric and Geometric distributions.
+
+///  Geometric Distribution
+/// Generates a sample from a geometric distribution with the given success probability.
+
+#[derive(Debug, PartialEq)]
+pub struct Geometric {
+    bernoulli: Bernoulli,
+}
+
+impl Geometric {
+    /// Creates a new `Geometric` distribution with the given success probability.
+    pub fn new(probability: f64) -> Result<Self, Error> {
+        if probability < f64::MIN_POSITIVE {
+            return Err(Error::BadGeometricProb(probability));
+        }
+        Ok(Self {
+            bernoulli: Bernoulli::new(probability)?,
+        })
+    }
+}
+
+impl Distribution<u32> for Geometric {
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> u32 {
+        let mut attempts = 0;
+        while !self.bernoulli.sample(rng) {
+            attempts += 1;
+        }
+        attempts
+    }
+}
+
+/// Double Geometric
+#[derive(Debug, PartialEq)]
+pub struct DoubleGeometric {
+    shift: u32,
+    geometric: Geometric,
+}
+
+impl DoubleGeometric {
+    /// Creates a new `DoubleGeometric` distribution with the given success probability and shift parameter.
+    pub fn new(s: f64, shift: u32) -> Result<Self, Error> {
+        if s < f64::MIN_POSITIVE {
+            return Err(Error::BadS(s));
+        }
+        if shift > 1_000_000 {
+            return Err(Error::BadSensitivity(shift));
+        }
+        let success_probability = 1.0 - E.powf(-1.0 / s);
+        Ok(Self {
+            shift,
+            geometric: Geometric::new(success_probability)?,
+        })
+    }
+}
+
+impl Distribution<i32> for DoubleGeometric {
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> i32 {
+        let attempts1 = self.geometric.sample(rng);
+        let attempts2 = self.geometric.sample(rng);
+        <u32 as TryInto<i32>>::try_into(self.shift).unwrap()
+            + <u32 as TryInto<i32>>::try_into(attempts1).unwrap()
+            - <u32 as TryInto<i32>>::try_into(attempts2).unwrap()
```

Comment:
that's an interesting idea.  I'd have to think a bit more how to prove what this ends up being but I plotted it and it doesn't look the same.  Even if I let in this way `self.geometric.sample(rng)` have a success probably that is half as much as the success probability in each of X_1 - X_2. 

Here's a python to look at the differences of the two sample distributions: 

`import numpy as np
import random


p1 = 0.3
p2 = 0.15
numsamples = 1000000

x_1 = np.random.geometric(p1,numsamples)
x_2 = np.random.geometric(p1,numsamples)

# compute first as X_1 - X_2
first = x_1 - x_2

firstcounts = {}
for d in first:
    if d in firstcounts:
        firstcounts[d] +=1
    else:
        firstcounts[d] = 1


# compute second as X_3 * (1 - 2 * Z)
second = []
for i in range(numsamples):
    z = bool(random.getrandbits(1))
    x_3 = np.random.geometric(p2)

    second.append(x_3 * (1 - 2 * z))

secondcounts = {}
for d in second:
    if d in secondcounts:
        secondcounts[d] +=1
    else:
        secondcounts[d] = 1

bound = 80
for i in range(-bound,bound+1):
    if i not in firstcounts:
        firstcounts[i] = 0
    if i not in secondcounts:
        secondcounts[i] = 0

for i in range(-bound,bound+1):
    # print(i,firstcounts[i],secondcounts[i],firstcounts[i] - secondcounts[i])
    print(i,firstcounts[i] - secondcounts[i])
`

Reply:
you can see the difference is not random but structured. 
![image](https://github.com/private-attribution/ipa/assets/35273659/5bbd9d65-eecd-4def-a496-22f4dba2dd7c)

Reply:
Interesting result.  I had a little trouble working out what the concrete differences were here.  Because your difference of geometric distributions is not a symmetric version of the same.  And I couldn't work out if this was really a discrete Laplace or not either.

Reply:
yeah, it it interesting. I have a reference that `X_1 - X_2` for iid geometrics will have an equivalent probability density to a discrete Laplace, but there the proof is left as an exercise :)  Be nice to understand but I haven't sat down to try and prove it directly.  I created a test that the sample distribution of this `X_1 - X_2` matches the expected probability distribution.